xvec/accessor.py

```diff
@@ -918,6 +929,128 @@ def to_geodataframe(
+        geometry_array = geo_series.geometry.array
+
+        xar_chunk = self._obj[var]
+        mask = rasterio.features.geometry_mask(geometry_array, out_shape=(xar_chunk.shape[0], xar_chunk.shape[1]), transform= trans)
+        masked_data = xar_chunk * mask[:, :, np.newaxis]
+        del mask, xar_chunk; gc.collect()
+
+        if stat == 'sum':
+            stat_within_polygons = da.sum(masked_data, axis=(0, 1))
+        elif stat == 'mean':
+            stat_within_polygons = da.mean(masked_data, axis=(0, 1))
+        elif stat == 'median':
+            stat_within_polygons = da.median(masked_data, axis=(0, 1))
+        elif stat == 'max':
+            stat_within_polygons = da.max(masked_data, axis=(0, 1))
+        elif stat == 'min':
+            stat_within_polygons = da.min(masked_data, axis=(0, 1))
+
+        result = stat_within_polygons.compute()
+        del masked_data, stat_within_polygons; gc.collect()
+
+        return result 
+    
+
+
+    def spatial_agg(self, geometries, stat='mean', chunk_size = 2):
+        transform = self._obj.rio.transform()
+        num_cores = multiprocessing.cpu_count() 
+        geometry_chunks = [geometries[i:i + chunk_size] for i in range(0, len(geometries), chunk_size)]
+
+        stats_dic = {}
+        for var in self._obj.data_vars:
+            stats_dic[var] = []
+
+            computed_results = []
+            for chunk in tqdm(geometry_chunks):
+                # Create a list of delayed objects for the current chunk
+                chunk_results =  Parallel(n_jobs=num_cores)(
+                    delayed(self.agg_geom)(geom, transform, var,  stat=stat) for geom in chunk
+                )
+                computed_results.extend(chunk_results)
+            stats_dic[var] = computed_results
+
+            gc.collect()
+
+        # Unpack the results into VectorCube
+        df = pd.DataFrame()
+        keys_items = {}
+        for k in stats_dic.keys():
+
+            s = stats_dic[k]
+            columns = []
+            for t in range(len(self._obj.time)):
+                columns.append(f'{k}{t+1}')
+            keys_items[k] = columns    
+            # Create a new DataFrame with the current data and columns
+            df_k = pd.DataFrame(s, columns=columns)
+            # Concatenate the new DataFrame with the existing DataFrame
+            df = pd.concat([df, df_k], axis=1)
+
+
+        #df = gpd.GeoDataFrame(df, geometry=gdf.geometry)
+        df = gpd.GeoDataFrame(df, geometry=geometries)
+        times = list(self._obj.time.values)
+
+        data_vars = {}
+        for key in keys_items.keys():
+            data_vars[key] = (["geometry", "time"], df[keys_items[key]])
+
+        ## Create VectorCube    
+        vec_cube = xr.Dataset(data_vars=data_vars, coords=dict(geometry=df.geometry, time=times)
+                         ).xvec.set_geom_indexes("geometry", crs=df.crs) 
+
+
+        return vec_cube
+
+    
+    def agg_polys(
```

Comment:
we should pick a better name. what is `{stars}` using?

Reply:
In R, `aggregate` is a generic, `{stars}` has a [method for stars objects](https://r-spatial.github.io/stars/reference/aggregate.stars.html)

Reply:
Both `rasterstats` and `xarray_spatial` use `zonal_stats,` so it may be best to use the same?

Reply:
Renamed to zonal_stats.